src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Vector128.cs

```diff
@@ -3032,6 +3032,37 @@ internal static void SetUpperUnsafe<T>(in this Vector128<T> vector, Vector64<T>
             Unsafe.AsRef(in vector._upper) = value;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        [CompExactlyDependsOn(typeof(AdvSimd.Arm64))]
+        internal static int CountMatches<T>(this Vector128<T> vector)
```

Comment:
> Given that these are perf critical APIs where the JIT can do a better thing for inputs in a particular shape, they should likely be JIT intrinsics so we get the best performance on ARm64.

Yes but that's not trivial and way above my skill level 😅 There is, however, a tracking issue for recognizing and optimizing movemask shape for ARM64: https://github.com/dotnet/runtime/issues/76047 But since there is more than one shape (with risk of recognition being fragile), I think this change is in the right direction. Also, writing movemask expecting the JIT to correctly rewrite it given set conditions for ARM64 may seem a bit counterintuitive.

> We also know we want such APIs, so we should really be taking these to API review so the general shape can be vetted first. For example, IndexOfMatch isn't sufficient and needs to be IndexOfFirstMatch and IndexOfLastMatch. Likewise CountMatches should likely be GetMatchCount() to be more inline with other similar Get*Count() names we have throughout the BCL.

Yes, my plan was to introduce internal helpers first, change existing SIMD paths to use them, and then create a proposal which can showcase profit/need in having these methods. Noted on naming - will fix.